ROPC: include for secure auth flow - managed identities (#33408)
dotnet/AspNetCore.Docs: aspnetcore/includes/managed-identities-test-non-production.md

---
author: wpickett
ms.author: wpickett
ms.date: 08/22/2024
ms.topic: include
---

> > [!NOTE]
> > Microsoft recommends that you use the most secure authentication flow available. The authentication flow described in this procedure is only suitable for testing and local development.
> >
> > Azure SQL Database should use [Managed Identities for Azure resources](/sql/connect/ado-net/sql/azure-active-directory-authentication#using-managed-identity-authentication) instead of the flow described in this tutorial. For non-Azure apps, use a secure authentication flow similar to managed identities for Azure resources.

Reviewer: ```suggestion
> [!NOTE]
> In this tutorial, a local database is used that doesn't use a password. For production apps:
> * A password is required.
> * Microsoft recommends that you use the most secure authentication flow available.
>
> Azure SQL Database should use [Managed Identities for Azure resources](/sql/connect/ado-net/sql/azure-active-directory-authentication#using-managed-identity-authentication). For non-Azure apps, use a secure authentication flow similar to managed identities for Azure resources.
>
> When the app is deployed to a test server, an environment variable can be used to set the connection string to a test database server. For more information, see [Configuration](xref:fundamentals/configuration/index).
```

---
author: wpickett
ms.author: wpickett
ms.date: 08/22/2024
ms.topic: include
---
> [!NOTE]

> > In this tutorial, a local database is used that doesn't use a password. For production apps:
> > * A password is required.
> > * Microsoft recommends that you use the most secure authentication flow available.
> >
> > Azure SQL Database should use [Managed Identities for Azure resources](/sql/connect/ado-net/sql/azure-active-directory-authentication#using-managed-identity-authentication). For non-Azure apps, use a secure authentication flow similar to managed identities for Azure resources.
> >
> > When the app is deployed to a test server, an environment variable can be used to set the connection string to a test database server. For more information, see [Configuration](xref:fundamentals/configuration/index).

Reviewer: Sorry I keep refining what I think we need. I don't think we need too much warning as this is an onboarding tutorial where almost everyone is using a local db.

```suggestion
> This tutorial uses a local database without a password. Production apps should use the most secure authentication flow available. For more information on authentication for deployed test and production apps , see [Secure authentication flows](/aspnet/core/security/<!-- write create article something like secureAuthn-->).
```
@tdykstra @wadepickett  what do you think?

Author: I appreciate the help and the thought you are putting into this, thanks.

Would this existing topic possibly work to point to?  It does a good job of helping to establish which scenario applies and what to look into from there for auth flows: 
https://learn.microsoft.com/en-us/entra/identity-platform/authentication-flows-app-scenarios

If not, would it at least work for now until a better topic is created?

Reviewer: > I appreciate the help and the thought you are putting into this, thanks.
> 
> I think we are back to the last one Tom updated of yours for now, until we have a better topic written to point to and then we can update the include in one place. That sound good?

Great. That's the power of includes.

Author: OK, I will not commit your shorter version above then, which would be great to use once we have a new topic that goes  into those details instead.  Thanks.

Reviewer 2: I would tweak the intro to the include a bit:

```suggestion
This article uses a local database that doesn't require the user to be authenticated. Production apps should use the most secure authentication flow available. For more information on authentication for deployed test and production apps, see [Secure authentication flows](/aspnet/core/security/<!-- write create article something like secureAuthn-->).
```

Author: Thanks @tdykstra . That is better and improves the last commit suggestion for a shorter one from Rick we are using sometime in the future when there is a new article that would include what is needed so we don't have to be wordy as above.  That would be for the future when we have that article.

It is this longer version (what is actually committed, example below) I think we are going with for now.  Is this below OK as is for now?

> In this tutorial, a local database is used that doesn't use a password. For production apps:
> * A password is required.
> * Microsoft recommends that you use the most secure authentication flow available.
>
> Azure SQL Database should use [Managed Identities for Azure resources](/sql/connect/ado-net/sql/azure-active-directory-authentication#using-managed-identity-authentication). For non-Azure apps, use a secure authentication flow similar to managed identities for Azure resources.
>
> When the app is deployed to a test server, an environment variable can be used to set the connection string to a test database server. For more information, see [Configuration](xref:fundamentals/configuration/index).

Author: If so and there is a review approval for using that last commited longer include for now I will merge.  Thanks again for all the help.

> [!NOTE]

> > Microsoft recommends that you use the most secure authentication flow available. The authentication flow described in this procedure requires a very high degree of trust in the application, and carries risks that are not present in other flows. You should only use this flow when other more secure flows, such as managed identities, aren't viable.

Reviewer: ```suggestion
<!-- we know what the most secure is, so make it a link and don't be so nebuluous and send them off hunting for the solution. -->
> Microsoft recommends that you use the [most secure authentication flow available](appropriate URL). The authentication flow described in this procedure:

* Is only suitable for test apps.
* Requires a very high degree of trust in the app.
* Carries risks that are not present in other flows.

This authentication flow should only be used in test apps. For production apps, use the most secure authentication flow available. <!-- TODO, links to recommended authentication -->
```

Author: Thanks @Rick-Anderson! Sorry, I don't know what the most secure authentication flow available is that we are now recommending now for this situation that I would link to.  We are assuming either SQL server or SQLite in this tutorial and more than one environment, but not assuming Azure nor an Azure account.

Reviewer: > Thanks @Rick-Anderson! Sorry, I don't know what the most secure authentication flow available is that we are recommending now for this situation that I would link to. Using this tutorial I am linking it to as an example, we are assuming either SQL server or SQLite in this tutorial and more than one environment, but not assuming Azure nor an Azure account.

My suggestion fixed

> [!NOTE]
> Microsoft recommends that you use the most secure authentication flow available. The authentication flow described in this procedure is only suitable for testing and local development.
>

> > Azure SQL Database should use [Managed Identities for Azure resources](/sql/connect/ado-net/sql/azure-active-directory-authentication#using-managed-identity-authentication) instead of the flow described in this tutorial. For non-Azure apps, use a secure authentication flow similar to [Managed Identities for Azure resources](/sql/connect/ado-net/sql/azure-active-directory-authentication#using-managed-identity-authentication).

Reviewer 2: ```suggestion
> Azure SQL Database should use [Managed Identities for Azure resources](/sql/connect/ado-net/sql/azure-active-directory-authentication#using-managed-identity-authentication) instead of the flow described in this tutorial. For non-Azure apps, use a secure authentication flow similar to managed identities for Azure resources.
```

---
author: wpickett
ms.author: wpickett
ms.date: 08/22/2024
ms.topic: include
---
> [!NOTE]

> > This article uses a local database that doesn't require the user to be authenticated. Production apps should use the most secure authentication flow available. For more information on authentication for deployed test and production apps, see [Secure authentication flows](/aspnet/core/security/<!-- write create article something like secureAuthn-->).

Reviewer: @wadepickett  create an issue to create `[Secure authentication flows](/aspnet/core/security/<!-- write create article something like secureAuthn-->).`

Author: Just to make sure I am in sync with the intent: 

- You added the commit that refers to a future topic not created so it is in the commit history along with Tom's last suggestion for it.
- Then you want me to pave over that with the last longer version of the include to merge today?

Author: Tracking issue: #33426 Secure Authentication Flows
